raiden_contracts/data/source/services/ServiceRegistry.sol

```diff
@@ -3,49 +3,167 @@ pragma solidity 0.5.4;
+        return true;
+    }
+}
+
+
 contract ServiceRegistry is Utils {
     Token public token;
-    address public owner;
 
-    mapping(address => uint256) public deposits;  // token amount staked by the service provider
+    // After a price is set to set_price at timestamp set_price_at,
+    // the price decays according to decayed_price().
+    uint256 public set_price;
+    uint256 public set_price_at;
+
+    // Once the price is at min_price, it can't decay further.
+    uint256 constant min_price = 1000;
+
+    mapping(address => uint256) public service_valid_till;
     mapping(address => string) public urls;  // URLs of services for HTTP access
-    address[] public service_addresses;  // list of available services (ethereum addresses)
+
+    event RegisteredService(address indexed service, uint256 valid_till, uint256 deposit_amount, Deposit deposit_contract);
 
     // @param _token_for_registration The address of the ERC20 token contract that services use for registration fees
-    constructor(address _token_for_registration) public {
+    constructor(address _token_for_registration, uint256 _initial_price) public {
         require(_token_for_registration != address(0x0), "token at address zero");
         require(contractExists(_token_for_registration), "token has no code");
+        require(_initial_price >= min_price, "initial price too low");
 
         token = Token(_token_for_registration);
         // Check if the contract is indeed a token contract
         require(token.totalSupply() > 0, "total supply zero");
-        owner = msg.sender;
-    }
-
-    function deposit(uint amount) public {
-        require(amount > 0);
 
-        // This also allows for MSs to deposit and use other MSs
-        deposits[msg.sender] += amount;
+        // Set up the price and the set price timestamp
+        set_price = _initial_price;
+        set_price_at = now;
+    }
 
-        // Transfer the deposit to the smart contract
-        require(token.transferFrom(msg.sender, address(this), amount), "tokens did not transfer");
+    // @notice Locks tokens and registers a service or extends the registration.
+    // @param _limit_amount The biggest amount of tokens that the caller is willing to deposit.
+    // The call fails if the current price is higher (this is always possible
+    // when other parties have just called `deposit()`).
+    function deposit(uint _limit_amount) public {
+        uint256 amount = current_price();
+        require(_limit_amount >= amount, "not enough limit");
+
+        // Extend the service position.
+        uint256 valid_till = service_valid_till[msg.sender];
+        if (valid_till < now) { // a first time joiner or an expired service.
+            valid_till = now;
+        }
+        // Check against overflow.
+        require(valid_till < valid_till + 180 days);
+        valid_till = valid_till + 180 days;
+        assert(valid_till > service_valid_till[msg.sender]);
```

Comment:
I'm not sure about low-level differences, but why do you use `assert`  here?

Reply:
Because I know the assert is always true.  So I use `assert` and then the SMT solver in `solc` tries to find a counterexample.

Reply:
To say it in a different way: `assert` should only fail for bugs in the contract, `require` is for all other cases.